shared/src/main/scala/mlscript/TypeDefs.scala

```diff
@@ -555,89 +555,92 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
+            updateVariance(lhs, curVariance.flip)
+            updateVariance(rhs, curVariance)
+          case Without(base, names) => updateVariance(base, curVariance.flip)
         }
-      case FunctionType(lhs, rhs) =>
-        updateVariance(lhs, curVariance.flip(VarianceInfo.contra()))
-        updateVariance(rhs, curVariance.flip(VarianceInfo.co()))
-      case Without(base, names) => updateVariance(base, curVariance.flip(VarianceInfo.contra()))
-    }
+      } (_ => if (varianceUpdated) s"updated - ${tyDef.nme.name}.${tyDef.tvarVariances})" else "end")
 
     do {
       varianceUpdated = false;
       tyDefs.foreach{
         case t@TypeDef(Cls, nme, _, _, body, mthDecls, mthDefs, _, _) =>
-          updateVariance(body, VarianceInfo.bi())(ctx, t)
+          updateVariance(body, VarianceInfo.co)(ctx, t)
           val stores = mthDecls.foreach { mthDef => 
             val mthBody = ctx.mthEnv.getOrElse(
               Right(Some(nme.name), mthDef.nme.name),
               throw new Exception(s"Method ${mthDef.nme.name} does not exist in the context")
             ).body
-            mthBody.foreach { case (_, body) => updateVariance(body, VarianceInfo.co())(ctx, t) }
+            mthBody.foreach { case (_, body) => updateVariance(body, VarianceInfo.co)(ctx, t) }
           }
         case t@TypeDef(Trt, nme, _, _, body, mthDecls, mthDefs, _, _) =>
-          updateVariance(body, VarianceInfo.bi())(ctx, t)
+          updateVariance(body, VarianceInfo.co)(ctx, t)
           val stores = mthDecls.foreach { mthDef => 
             val mthBody = ctx.mthEnv.getOrElse(
               Right(Some(nme.name), mthDef.nme.name),
               throw new Exception(s"Method ${mthDef.nme.name} does not exist in the context")
             ).body
-            mthBody.foreach { case (_, body) => updateVariance(body, VarianceInfo.co())(ctx, t) }
+            mthBody.foreach { case (_, body) => updateVariance(body, VarianceInfo.co)(ctx, t) }
           }
         case t@TypeDef(Als, nme, _, _, body, mthDecls, mthDefs, _, _) =>
-          updateVariance(body, VarianceInfo.bi())(ctx, t)
+          updateVariance(body, VarianceInfo.co)(ctx, t)
       }
     } while (varianceUpdated)
   }
 
   case class VarianceInfo(isCovariant: Bool, isContravariant: Bool) {
 
+    /** Combine two pieces of variance information together
+     */
     def &&(that: VarianceInfo): VarianceInfo =
       VarianceInfo(isCovariant && that.isCovariant, isContravariant && that.isContravariant)
 
-    def flip(that: VarianceInfo): VarianceInfo = that match {
-      case (VarianceInfo(true, true)) => this
-      case (VarianceInfo(true, false)) => if (this === VarianceInfo.bi()) that else this
-      case (VarianceInfo(false, true)) => if (this === VarianceInfo.bi()) that else VarianceInfo(isContravariant, isCovariant)
-      case (VarianceInfo(false, false)) => this
-    }
+    /*  Flips the current variance if it encounters a contravariant position
```

Comment:
For consistency with the doc for `&&`:
```suggestion
    /*  Flip the current variance when going through a contravariant position
```